src/bin/cargo/cli.rs

```diff
@@ -222,16 +219,30 @@ fn add_ssl(version_string: &mut String) {
     }
 }
 
+/// Expands aliases recursively to collect all the command line arguments.
+///
+/// [`GlobalArgs`] need to be extracted before expanding aliases because the
+/// clap code for extracting a subcommand discards global options
+/// (appearing before the subcommand).
 fn expand_aliases(
     config: &mut Config,
     args: ArgMatches,
     mut already_expanded: Vec<String>,
 ) -> Result<(ArgMatches, GlobalArgs), CliError> {
     if let Some((cmd, args)) = args.subcommand() {
-        match (
-            commands::builtin_exec(cmd),
-            super::aliased_command(config, cmd)?,
-        ) {
+        let exec = commands::builtin_exec(cmd);
+        let aliased_cmd = super::aliased_command(config, cmd).or_else(|e| {
+            // HACK: `cargo version` must always work even with malformed configs.
+            // Here we treat the error as the aliased command not found,
+            // so it falls back to call builtin `cargo version`
+            if cmd == "version" {
```

Comment:
> In addition, cargo version must work even with malformed configurations. I've [hardcoded version str](https://github.com/rust-lang/cargo/pull/11087/commits/41ca806c4324ef39bfb580d8b4b4281fe34ea613#diff-86f2b77af0be4f1b5c555ec44a7aa6f04d5bd797bf7f5667ee4f9c56e042f11eR235-R242) to do dirty works. It's not ideal, and any suggestion is welcome.

What if we removed the `?`?
- The warning case would be updated from `(Some(_), Some(_))` to `(Some(_), Ok(Some(_)))`
- The built-in case would be updated from `(Some(_), None)` to `(Some(_), _)`
- We would then add a `(None, Err(err))` case for reporting the error

Since we prioritize built-in anyways, this makes **all** built-ins work if the config is bad but then reports the error when aliases are being used.

Reply:
> What if we removed the `?`?

It is a good suggestion for readability! However we cannot do that because doing that Cargo will lose the warning of _"trailing arguments after built-in command \`version\` are ignored: \`-v\`"_ when running `cargo -- version -v` with a bad config.

> Since we prioritize built-in anyways, this makes **all** built-ins work if the config is bad but then reports the error when aliases are being used.

By built-in, are you referring to built-in alias or built-in command? Seems like Cargo [doesn't favour built-in aliases](https://github.com/rust-lang/cargo/blob/aaadab6db35be307c0548be3b299bf0a045e7025/src/bin/cargo/main.rs#L69-L71) over user-defined ones 🥲. Do you have an example at hand that we can improve its error message?

Reply:
> However we cannot do that because doing that Cargo will lose the warning of "trailing arguments after built-in command `version` are ignored: `-v`" when running cargo -- version -v with a bad config.

Good timing for me getting back to this.  That is now an error (https://github.com/rust-lang/cargo/pull/11119) and its just a question of which error to prioritize.

Reply:
> By built-in, are you referring to built-in alias or built-in command? 

built-in commands (`commands::builtin_exec`)

Reply:
I prioritize alias resolution over trailing arguments with 903783a.